### ISSUE TITLE
performance of SqlTileWriter.runCleanupOperation()

### ISSUE THREAD

**Issue by shiroyagi** [osmdroid/osmdroid#965]
## Issue Type

[ *] Bug

## Description and/or steps/code to reproduce the problem

When the maximum cache size is exceeded and runCleanupOperation() is called, it can cause the device to freeze for several minutes.  Presumably this is a very cpu and disk intensive operation and can overwhelm the device to such an extent that it stops responding and I can't close the application.  I do have cleanOnStartup=false to avoid purging the cache if I don't have network connection.

Possible ideas for improvement:

1. create an index for COLUMN_EXPIRES
2. perform 'DELETE FROM' in several chunks (eg. 10) rather than one go (perhaps not without an index?)
3. split the database into several (eg. 5) files and just delete the oldest one when the cache size is exceeded (may slow tile search?)
4. don't batch delete tiles at all, just delete/replace individual tiles with newly downloaded ones when full

Maybe some of these have been tried before, or are known not to work?

### If it's a bug, version(s) of android this affects:

5.1

### Version of osmdroid the issue relates to:

6.0.1

**Comment by monsieurtanuki** [comment]
`SqlTileWriter.runCleanupOperation` is commented: _this could be a long running operation, don't run on the UI thread unless necessary_.

It's currently called in 3 places :
* in the `cleanOnStartup` mode, it's called async'ly in a `Thread` with low priority
* when it's called in `saveFile`, it's called sync'ly in two cases: beyond the functional space threshold, or when there's really no space left.

The code itself is:
* check database size
* if it's not too big (by default, under 600Mb), return
* pre-computation with unnecessary approximation, and then a big ineffective SQL `DELETE` statement

In your case, the problems I see are:
* it's called sync'ly
* it's called with a very bad timing - as you're currently saving a new tile into the database
* it's _very_ slow

What I suggest in order to fix it:
* call it async'ly (instead of sync'ly) when you exceed the threshold (when it's first called in `saveFile`), possibly using a singleton in order to avoid running tons of `Thread`s
* as you suggested, split the big `DELETE` statement into a loop of single row `DELETE` statements - something like "while we are beyond the threshold, get the eldest row, delete the row, continue if needed". No more bulk `DELETE`, just a string of harmless single row `DELETE` statements

To be confirmed: I guess the main problem with the current big `DELETE` statement is the eldest tiles selection as there's no index: it must perform a full scan of the SQL table.
Of course adding an SQL index on column `expires` would help,  but it takes space in the database too.

I also have this dream of splitting the database into 30 tables, one for each zoom level (0-29). Doing so you would have better performances (smaller tables), and you'd be able to run SQL statements in parallel on all tables. Maybe not 100% relevant in this issue, though.

**Comment by shiroyagi** [comment]
I think the biggest problem is that it is very slow and intensive, it froze the phone not just the app so I don't think another thread or timing is going to help much until that is resolved.

Is a big DELETE necessarily ineffective?  I could imagine that there are efficiencies to be gained by reducing the number of queries/calls, particularly if it has to search for the oldest expiry each time.  I was thinking 10 not quite so big deletes, not thousands.

Adding an index also has potential to make a big delete slower, as it will have to be rebuilt in that case.

Perhaps ideas 3 (delete whole file outside of SQL) & 4+1 (delete individually on demand using index) have the greatest potential.  It can be quite laborious to test since it need a reasonable size cache to be noticeable, which takes a lot of map viewing to fill.  In normal usage there are a few months between these events.  However for these two solutions the problem is eliminated by design and it is easier to test that general cache performance is not significantly affected.

**Comment by spyhunter99** [comment]
good points all around. We can add indexes to the expires column if needed, that may be the easiest

**Comment by monsieurtanuki** [comment]
Additional comments (regardless of the specific recycling algorithm):
* I'd like to implement some kind of "garbage collector" design pattern (not sure if that exists), with an async `gc` singleton method
* we also need a unit test class

**Comment by shiroyagi** [comment]
`SqlTileWriter.getFirstExpiry()` currently runs in about 25s whose query is part of the current cleanup and too slow to be done frequently.  I will first check what impact an index has on this time, and then try tuning the ratio `getTileFileSystemCacheTrimBytes()/getTileFileSystemCacheMaxBytes()`.  The default is 0.83, as it approaches 1.0 it should be equivalent to a 'one in, one out' policy.  Hopefully some combination of these will give good results, and the code can then be optimised.

**Comment by shiroyagi** [comment]
Adding an index results in virtually instantaneous `SqlTileWriter.getFirstExpiry()`:
`db.execSQL("CREATE INDEX IF NOT EXISTS " + COLUMN_EXPIRES_INDEX + " ON "+ TABLE + " (" + COLUMN_EXPIRES +");");`
It takes around 25s to build the index on first use, but this speed of access suggests a trim ratio of close to 1.0 is practical (currently testing with 0.95).

**Comment by InI4** [comment]
Just a loose idea: If this was small steps/async thread, that thread could kill tiles proactively, i.e. when a certain second slightly lower ratio is reached, tiles are removed from database before the final memory limit, i.e. 1.0 is actually hit. This could be done with low priority and sleeps in between and would avoid things getting slow, when the user actually interacts.

**Comment by shiroyagi** [comment]
Yes, and async threaded gc sounds good as long as it doesn't block reading/writing tiles to the database.  So it could start looping over single tile deletes after a few seconds of inactivity following tiles being added.  Any subsequent read/write could then interrupt the loop.

As a first step, I can try such a loop in the main thread immediately after adding a tile.

**Comment by shiroyagi** [comment]
There is a potential problem, in runCleanupOperation():
```
                //note, i considered adding a looping mechanism here but sqlite can behave differently
                //i.e. there's no guarantee that the database file size shrinks immediately.
```
ie. either have to keep track of the size of tiles deleted, or using the average tile size just estimate how many to remove hoping they will balance out in the long term.  There is wide variety in tile sizes (a factor of 7 or more).  Given the performance requirements I would set the DELETE LIMIT to `newTileSize/averageTileSize`.

**Comment by shiroyagi** [comment]
Some further testing.  I have just these changes compared to default 6.0.1:
1. expired index in database
2. trimBytes==maxBytes
3. saveFile() clean up delay from lastSizeCheck is 1s rather than 5min

The cache stays permanently at close to 100% full and tile download performance is not noticeably affected.

```
03-19 22:29:02.750 D/OsmDroid( 3223): Local cache purging 7 tiles.
03-19 22:29:03.078 D/OsmDroid( 3223): purge completed in 328ms, cache size is 629456896 bytes
03-19 22:29:07.849 D/OsmDroid( 3223): Local cache purging 8 tiles.
03-19 22:29:07.911 D/OsmDroid( 3223): purge completed in 62ms, cache size is 629501952 bytes
03-19 22:29:08.909 D/OsmDroid( 3223): Local cache purging 10 tiles.
03-19 22:29:08.963 D/OsmDroid( 3223): purge completed in 53ms, cache size is 629596160 bytes
03-19 22:29:10.580 D/OsmDroid( 3223): Local cache purging 13 tiles.
03-19 22:29:10.647 D/OsmDroid( 3223): purge completed in 66ms, cache size is 629682176 bytes
03-19 22:29:18.937 D/OsmDroid( 3223): Local cache purging 22 tiles.
03-19 22:29:19.233 D/OsmDroid( 3223): purge completed in 296ms, cache size is 629510144 bytes
03-19 22:29:20.072 D/OsmDroid( 3223): Local cache purging 11 tiles.
03-19 22:29:20.587 D/OsmDroid( 3223): purge completed in 515ms, cache size is 629460992 bytes
03-19 22:29:21.080 D/OsmDroid( 3223): Local cache purging 7 tiles.
03-19 22:29:21.401 D/OsmDroid( 3223): purge completed in 321ms, cache size is 629391360 bytes
03-19 22:29:22.172 D/OsmDroid( 3223): Local cache purging 7 tiles.
03-19 22:29:22.318 D/OsmDroid( 3223): purge completed in 146ms, cache size is 629334016 bytes
03-19 22:42:47.508 D/OsmDroid( 3223): Local cache purging 16 tiles.
03-19 22:42:48.099 D/OsmDroid( 3223): purge completed in 591ms, cache size is 629698560 bytes
03-19 22:42:48.517 D/OsmDroid( 3223): Local cache purging 17 tiles.
03-19 22:42:48.605 D/OsmDroid( 3223): purge completed in 88ms, cache size is 629841920 bytes
03-19 22:42:49.558 D/OsmDroid( 3223): Local cache purging 33 tiles.
03-19 22:42:49.709 D/OsmDroid( 3223): purge completed in 152ms, cache size is 630632448 bytes
03-19 22:42:50.781 D/OsmDroid( 3223): Local cache purging 45 tiles.
03-19 22:42:51.577 D/OsmDroid( 3223): purge completed in 796ms, cache size is 630538240 bytes
03-19 22:42:51.951 D/OsmDroid( 3223): Local cache purging 28 tiles.

```
The tile deletion times here are quite small and not too dissimilar to downloading from the server.  The later larger and slower purges correspond to downloading from a source with larger tile images.

This seems acceptable as a short term fix, but it would be nice if the deletion were concurrent with downloading.

Perhaps we could create an SQL transaction manager where read queries are executed synchronously with high priority, and add and delete asynchronously from a queue with low priority?  This would just require the corresponding methods to be synchronized on the manager instance.

**Comment by InI4** [comment]
Another random thought. @monsieurtanuki  talked about splitting the tile cache in 30 tables. When you use some module hacking one could restrict this to (say) 6 tables. The chance that concurrency happens between zoom level x and zoom level x+6 is significantly reduced, but you do not mess things up too much.

**Comment by shiroyagi** [comment]
I'm not sure you want concurrency between multiple SQL transactions since this may overload the system (cpu and I/O) and restore the original problem.  Splitting by expiry rather than zoom gives a low cost method of bulk delete.  Alternatively splitting by tile source would simplify determining cache size used by each source, and clearing them independently.

In the above idea for a transaction manager, you may not need any additional serialisation with a single database since it should already ensure this (ACID).

**Comment by monsieurtanuki** [comment]
@shiroyagi What's the next step? If I sum up, the main fix to this issue is creating the SQL index. Which can be problematic for the very first time we create the index on old databases, as it can take dozens of seconds and it's currently run sync'ly.
All we need now for decent results is an asynchronous call to "first time only index creation + possible eldest records deletion".
I'm currently working on a class that would make those frequent async calls possible; would that help close this issue?

**Comment by shiroyagi** [comment]
Yes the index is key, but also I think doing the deletes in relatively small but frequent chunks.  Therefore I would change the default to trimBytes==maxBytes and reduce `lastSizeCheck`, possibly making the latter a `Configuration` or static member option to allow tuning.

For the index creation, I'm assuming any other database operations (tile access) will block or be very slow during this time.  So ideally it would be asynchronous, but prevent other calls from blocking the UI thread.  How easy this is, and whether it is worth doing for a one-off is another question.  I'd be tempted just to show a Toast before it starts to warn what is going on, or even just a warning in the release notes?

In terms of good ongoing performance, I think a queue of tile addition/removal transactions to be run asynchronously has the most potential.  Is this what your class will do?  There may be some subtleties such as keeping track of the bytes delta in the queue, and the previously noted lag between database size and file size (SQLiteDatabase caching?) in terms of maintaining the desired file size.  This could be left to a future optimisation task: my simple changes noted above seem to be good enough on my hardware.

Thanks for your effort, I haven't been able to spend too much time on this recently.

**Comment by monsieurtanuki** [comment]
@shiroyagi My idea is:
* each time we save a tile into SQL, we run a custom `gc`
* this `gc` can only be run one at a time
* this is async
* the code is "create the index if not exists, check the db size and delete the eldest records up to the limit"

It will be smooth this way.
There's still the problem of the SQL database not shrinking, but I cannot see an elegant way to know the accurate size of the db. Possibly in SQLite system tables?

**Comment by InI4** [comment]
From an abstract perspective: Tile handling comes in burst. An indexed SQlite database is usually more efficient on (medium) batch operations, so deleting (say) 8 tiles at a time, should generate even less load. If tiles are deleted in batches, it should be either create an index or delete a tile.

**Comment by monsieurtanuki** [comment]
Just PR'ed my fix in #973

**Comment by monsieurtanuki** [comment]
Supposedly fixed in #973.
@shiroyagi Feel like testing?

**Comment by monsieurtanuki** [comment]
@shiroyagi ping
OK to close this issue?

**Comment by monsieurtanuki** [comment]
@shiroyagi ping

**Comment by monsieurtanuki** [comment]
From what I read in  #1171, I assume that the garbage collector's performances are not longer an issue like the original several minute freeze described by the OP.
Feel free to reopen if needed.
